File: CodeSprint-2012-04-28-ExtremeStartups-Canada/MoneyGambling/mg.php
<?php

	function Pr($n, $i)
	{
		global $pr, $p, $k;

		$key = $n.'-'.$i;

		if (! array_key_exists($key, $pr))
		{
			if ($i > $n)
			{
				$pr[$key] = 0;
			}
			elseif ($n === 1)
			{
				if ($i === 0)
				{
					$pr[$key] = 1 - $p;
				}
				else
				{
					$pr[$key] = $p;
				}
			}
			elseif ($i === 0)
			{
				$pr[$key] = (1 - Pr0($n - 1)) * (1 - $p);
			}
			else
			{
				$pr[$key] = Pr($n - 1, $i - 1) * $p;
			}
		}

		return $pr[$key];
	}

	function Pr0($n)
	{
		global $pr0, $k;

		if (! array_key_exists($n, $pr0))
		{
			if ($n < $k)
			{
				$pr0[$n] = 0;
			}
			else
			{
				$pr0[$n] = Pr0($n - 1) + Pr($n, $k);
			}
		}

		return $pr0[$n];
	}

	$T = intval(readline(''));

	for (; $T; --$T)
	{
		list($p, $k) = explode(' ', readline(''), 2);
		$p = floatval($p);
		$k = intval($k);
	
		$pr = array();
		$pr0 = array();
	
		$prob = 0;
		$po = 0;
		$exp = 0;
		$exppo = 0;
	
		for ($i = 1; ; ++$i)
		{
			$po += (2 * $i) - 1;
	
			$pcur = Pr($i, $k);
	
			if ($i >= $k && ($pcur < 1e-7))
			{
				break;
			}
	
			$prob += $pcur;
			$exp += $i * $pcur;
			$exppo += $po * $pcur;
		}
	
		print(floor($exp).' '.floor($exppo)."\n");
	}

?>

File: CodeSprint-2012-04-28-ExtremeStartups-Canada/MoneyGambling/mg.cxx

#include <iostream>
#include <istream>
#include <ostream>

#include <cmath>

#include <map>
#include <vector>
#include <utility>

using namespace std;

long double Pr0(long long n, map<pair<long long, long long>, long double> &pr, map<long long, long double> &pr0, long double p, long long k);

long double Pr(long long n, long long i, map<pair<long long, long long>, long double> &pr, map<long long, long double> &pr0, long double p, long long k)
{
	pair<long long, long long> key(n, i);

	if (pr.find(key) == pr.end())
	{
		if (i > n)
		{
			pr[key] = 0;
		}
		else if (n == 1)
		{
			if (i == 0)
			{
				pr[key] = 1 - p;
			}
			else
			{
				pr[key] = p;
			}
		}
		else if (i == 0)
		{
			pr[key] = (1 - Pr0(n - 1, pr, pr0, p, k)) * (1 - p);
		}
		else
		{
			pr[key] = Pr(n - 1, i - 1, pr, pr0, p, k) * p;
		}
	}

	return pr[key];
}

long double Pr0(long long n, map<pair<long long, long long>, long double> &pr, map<long long, long double> &pr0, long double p, long long k)
{
	if (pr0.find(n) == pr0.end())
	{
		if (n < k)
		{
			pr0[n] = 0;
		}
		else
		{
			pr0[n] = Pr0(n - 1, pr, pr0, p, k) + Pr(n, k, pr, pr0, p, k);
		}
	}

	return pr0[n];
}

int main()
{
	int T;

	cin >> T;

	for (; T; --T)
	{
		long double p;
		long long k;

		cin >> p >> k;

		map<pair<long long, long long>, long double> pr;
		map<long long, long double> pr0;

		long long po = 0;
		long double exp = 0;
		long double exppo = 0;

		for (long long i = 1; ; ++i)
		{
			po += (2 * i) - 1;

			long double pcur = Pr(i, k, pr, pr0, p, k);

			if (i >= k && (pcur < 1e-12))
			{
				break;
			}

			exp += i * pcur;
			exppo += po * pcur;
		}

		cout << floor(exp) << ' ' << floor(exppo) << endl;
	}

	return 0;
}

